codelists/presenters.py

```diff
@@ -1,8 +1,72 @@
 from itertools import groupby
 
+import attr
 from django.urls import reverse
 
 
+@attr.s
+class Row:
+    """
+    Data structure for a Definition to prepare for display
+
+    name: name of the definition
+    code: code of the definition
+    excluded_children: list of excluded children (created with this class too)
+    all_descendants: are all of this Definitions descendants included?
+
+    Between all_descendants and excluded_children we can cover the three state
+    situation a Definition can be in:
+
+        * all descendants are included (all_descendants = True, excluded_children = [])
+        * all descendants except N (all_descendants = True, excluded_children = [...])
+        * no descendants (all_descendants = False, excluded_children ignored)
+
+    """
+
+    name: str = attr.ib()
+    code: str = attr.ib()
+    excluded_children: list = attr.ib(default=list())
+    all_descendants: bool = attr.ib(default=True)
+
+
+def _iter_rules(hierarchy, rules, name_for_rule, excluded=None):
+    for rule in rules:
+        row = Row(
+            name=name_for_rule(rule),
+            code=rule.code,
+            all_descendants=rule.applies_to_descendants,
+        )
+
+        # no descendents for this code so we can shortcut the iteration here
+        if not rule.applies_to_descendants:
+            yield attr.asdict(row)
+            continue
+
+        # filter the excluded rules down to children of the current Rule
+        excluding_rules = [
+            r for r in excluded if r.code in hierarchy.descendants(rule.code)
+        ]
+
+        # generate excluded children by recursing into this function
+        row.excluded_children = list(
+            _iter_rules(hierarchy, excluding_rules, name_for_rule)
+        )
```

Comment:
I found the recursion a bit hard to follow -- in part because on the first call, `rules` are included rules, and on the second call (ie the first recursion) they are excluded rules.

Given that on the first recursion, we only yield `Row(name=name_for_rule(rule), code=rule.code, all_descendants=False)`, I'd be tempted to avoid recursing altogether, and build up `row.excluded_children` directly.

Reply:
Great shout, this was very much a case of not seeing the wood for the trees.  Fixed in dad666a and much easier to read I think!